common/component/kafka/kafka.go

```diff
@@ -166,6 +223,128 @@ func (k *Kafka) Close() (err error) {
+
+		if len(message.Value) < 5 {
+			return nil, fmt.Errorf("value is too short")
+		}
+		schemaID := binary.BigEndian.Uint32(message.Value[1:5])
+		schema, err := srClient.GetSchema(int(schemaID))
+		if err != nil {
+			return nil, err
+		}
+		// The data coming through is standard JSON. The version currently supported by srclient doesn't support this yet
+		// Use this specific codec instead.
+		codec, err := goavro.NewCodecForStandardJSONFull(schema.Schema())
+		if err != nil {
+			return nil, err
+		}
+		native, _, err := codec.NativeFromBinary(message.Value[5:])
+		if err != nil {
+			return nil, err
+		}
+		value, err := codec.TextualFromNative(nil, native)
+		if err != nil {
+			return nil, err
+		}
+
+		return value, nil
+	}
+	return message.Value, nil
+}
+
+func (k *Kafka) getLatestSchema(topic string) (*srclient.Schema, *goavro.Codec, error) {
+	srClient, err := k.getSchemaRegistyClient()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	subject := getSchemaSubject(topic)
+	if k.schemaCachingEnabled {
+		cacheEntry, ok := k.latestSchemaCache[subject]
+		// Cache present and not expired
+		if ok && cacheEntry.expirationTime.Compare(time.Now()) < 0 {
+			return cacheEntry.schema, cacheEntry.codec, nil
+		}
+		schema, errSchema := srClient.GetLatestSchema(subject)
+		if errSchema != nil {
+			return nil, nil, errSchema
+		}
+		// New JSON standard serialization/Deserialization is not integrated in srclient yet.
+		// Since standard json is passed from dapr, it is needed.
+		codec, errCodec := goavro.NewCodecForStandardJSONFull(schema.Schema())
+		if errCodec != nil {
+			return nil, nil, err
+		}
+		k.latestSchemaCacheLock.Lock()
+		k.latestSchemaCache[subject] = SchemaCacheEntry{schema: schema, codec: codec, expirationTime: time.Now().Add(k.latestSchemaCacheTTL)}
+		k.latestSchemaCacheLock.Unlock()
+		return schema, codec, nil
+	}
+	schema, err := srClient.GetLatestSchema(getSchemaSubject(topic))
+	if err != nil {
+		return nil, nil, err
+	}
+	codec, err := goavro.NewCodecForStandardJSONFull(schema.Schema())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return schema, codec, nil
+}
+
+func (k *Kafka) getSchemaRegistyClient() (srclient.ISchemaRegistryClient, error) {
+	if k.srClient == nil {
+		return nil, errors.New("schema registry details not set")
+	}
+
+	return k.srClient, nil
+}
+
+func (k *Kafka) SerializeValue(topic string, data []byte, metadata map[string]string) ([]byte, error) {
+	valueSchemaType, err := GetValueSchemaType(metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	if valueSchemaType == Avro {
+		schema, codec, err := k.getLatestSchema(topic)
+		if err != nil {
+			return nil, err
+		}
+
+		native, _, err := codec.NativeFromTextual(data)
+		if err != nil {
+			return nil, err
+		}
+
+		valueBytes, err := codec.BinaryFromNative(nil, native)
+		if err != nil {
+			return nil, err
+		}
+		schemaIDBytes := make([]byte, 4)
+		binary.BigEndian.PutUint32(schemaIDBytes, uint32(schema.ID()))
+
+		var recordValue []byte
```

Comment:
Nit: optimize memory allocation (hopefully I got the right count)

```suggestion
		recordValue := make([]byte, 0, len(schemaIDBytes)+len(valueBytes)+1)
```